Brokerages/Backtesting/BacktestingBrokerage.cs

```diff
@@ -398,19 +398,29 @@ public void SimulateMarket()
         /// <param name="quantity">Quantity to assign</param>
         public virtual void ActivateOptionAssignment(Option option, int quantity)
         {
-            var request = new SubmitOrderRequest(OrderType.OptionExercise, option.Type, option.Symbol, -quantity, 0.0m, 0.0m, Algorithm.UtcTime, "Simulated option assignment");
-            var order = (OptionExerciseOrder)Order.CreateOrder(request);
-            var fills = option.OptionExerciseModel.OptionExercise(option, order);
-            var portfolioModel = (OptionPortfolioModel)option.PortfolioModel;
+            // do not process the same assignment more than once
+            if (_pendingOptionAssignments.Contains(option.Symbol)) return;
 
-            foreach (var fill in fills)
-            {
-                // processing assignment in the portfolio first
-                portfolioModel.ProcessFill(Algorithm.Portfolio, option, fill);
+            _pendingOptionAssignments.Add(option.Symbol);
+
+            var request = new SubmitOrderRequest(OrderType.OptionExercise, option.Type, option.Symbol, -quantity, 0m, 0m, Algorithm.UtcTime, "Simulated option assignment before expiration");
 
-                // firing event informing interested parties that option position has been assigned
-                OnOptionPositionAssigned(fill);
+            var ticket = Algorithm.Transactions.ProcessRequest(request);
+            Log.Trace($"BacktestingBrokerage.ActivateOptionAssignment(): OrderId: {ticket.OrderId}");
+        }
+
+        /// <summary>
+        /// Event invocator for the OrderFilled event
+        /// </summary>
+        /// <param name="e">The OrderEvent</param>
+        protected override void OnOrderEvent(OrderEvent e)
+        {
+            if (e.Status == OrderStatus.Filled && _pendingOptionAssignments.Contains(e.Symbol))
```

Comment:
Should we also be checking for canceled and such? Perhaps using e.Status.IsClosed() is more inclusive?

Reply:
This check was only added to prevent `_pendingOptionAssignments` from growing indefinitely, but you're right, the assignment could fail for insufficient margin and the order status could be `Invalid`. Agree `e.Status.IsClosed()` is a better choice.